text/3081-multi-part-examples-in-rustdoc/index.md

```diff
@@ -0,0 +1,233 @@
+  This proposal would add another layer of nesting to this.
+  Another layer of nesting would incur mental overhead and increased implementation cost for tooling compared with the suggested approach.
+- For longer nested comments, might be cumbersome.
+  Imagine comments that include code blocks. That's another nesting. Imagine long text. You'd have to handle line breaks and comment slashes (`//`).
+- [According to @jyn514](https://github.com/rust-lang/rfcs/pull/3081#discussion_r577914359), this would be very difficult to implement.
+
+Advantages:
+
+- Seems rather futuristic and perhaps could be part of a larger langauge change, where all comments are by definition markdown, and therefore GUI editors could display rendered versions of comments in place of the actual comments.
+- There's no splitting of code blocks.
+
+We imagine a more convenient authoring experience with the suggested design.
+Even though it feels like a close call, that tips the scale.
+
+Also, the suggested approach is not mutually exclusive with this approach, so that this approach could remain a future possibility.
+
+## Tree variant
+
+This design is similar to the suggested design, but blocks may form tree structures. Multiple blocks may continue from a same prior block.
+
+The advantage is that it would allow code sharing between multiple examples.
+
+The disadvantage is that the result might not be reader-friendly.
+Documentation is pages that are consumed usually top to bottom.
+Adding code blocks that continue one another in some tree structure may seem like a good idea from the perspective of the developer/author but would probably not actually be nice for the reader.
+
+## Interleaved variant
+
+This is a feature that is additional to the suggested design, where the blocks of a multi-part example may interleave with blocks of other multi-part examples.
+
+We feel that this could lead to confusing documentation.
+
+## Q: What is the impact of not doing this?
+
+If this feature is not implemented and no other solution is implemented for including text in long code examples, then long code examples in rust documentation continue to be inconvenient for both authors and readers.
+
+# Prior art
+[prior-art]: #prior-art
+
+This RFC is the evolution of a comment analyzing the existing [RFC 2894][rfc_2894]. It follows [this comment][rfc_comment].
+
+If we look at how programming books are commonly formatted, we will often find multi-part code examples.
+They are split into multiple code blocks and continue each other linearly.
+Between the code blocks is usually verbose educational text.
+Rust documentation authors already seem to follow this style while using one or more of the workarounds described earlier.
+This RFC officialy adopts this style.
+
+One could claim that this RFC allows code examples to be written in [literate programming][literate_programming] style.
+
+We are not aware of a similar feature in any other language.
```

Comment:
Python has a long history of using doctests: comments with embedded Python interpreter sessions. They look like this:

```rst
This is an example text file in reStructuredText format.  First import
``factorial`` from the ``example`` module:

    >>> from example import factorial

Now use it:

    >>> factorial(6)
    120
```

Notice how the two `>>>` lines share state: they're part of one interpreter session. The example is from the [`doctest` documentation](https://docs.python.org/3/library/doctest.html). This style of comment was supported at least [20 years ago as seen in PEP-287](https://peps.python.org/pep-0287/).

The Python doctest library has been ported to other languages, such as [Haskell](https://github.com/sol/doctest). The documentation for that library explains that it clears the state between each code block by default (but there is a `--fast` flag which disables this since there is a performance cost to clearing the state).

Reply:
Why don't we just concatenate code blocks within a single doc comment?
It seems like most code blocks within a doc comment are in the context of the item being documented.

```rust
/// We can bind a value to a variable
/// ```
/// let x = 0;
/// ```
/// We can use the bound variable as a value in another binding
/// ```
/// let y = x;
/// ```
/// We can use `shadowing` to bind a new variable with the name of an existing variable
/// ```
/// let x = 5;
/// ```
/// Note that the original variable `x` is still in scope, but can no longer be directly accessed
```

This would compile to:
```rust
let x = 0;
let y = x;
let x = 5;
```

For the most part, it seems variable shadowing can easily be used to soft-reset variable names without resetting the entire scope. And if users are really particular about their scope they can be explicit with braces (with or without #).
```rust
/// ```
/// # {
/// let x = 0;
/// ```
/// ```
/// let y = x;
/// # }
/// ```
/// ```
/// let x = 5;
/// ```
```

Are there prior discussions with reasoning for avoiding this?

Reply:
Thank you for the suggestion, @beeryt. Do you feel that implicit concatenation of code blocks fits in with other design decisions in Rust?

For one, that would be a breaking change.

Reply:
Combining by default would break `std` and quite a huge amount of crates. I don't think an attribute is the right way to do it either. Using codeblock labels seems like a better idea:

````rust
/// ```combine
/// let x = 0;
/// ```
///
/// blabla
///
/// ```
/// let x = 1;
/// ```
///
/// ```combine
/// assert!(x == 0);
/// ```
````

Reply:
I see it now, in `std::io::Read` for instance there are multiple definitions of main. I'm afraid variable shadowing doesn't apply to non-variables.

However, I still think an attribute (item-level *not* crate-level) makes sense since I envision it being used in scenarios where authors explicitly want concatenation behavior.

Of course, the attribute idea is incompatible with your example above. However, I'm not sure how useful it is when it could alternatively be expressed as:

```rust
#[doc(test(combine_code_blocks))]
/// ```
/// let x = 0;
/// ```
/// blabla
/// ```ignore (or maybe no_run)
/// let x = 1;
/// ```
/// ```
/// assert!(x == 0);
/// ```
```

Reply:
You mentioned yourself that an attribute would break std docs. So from that, I don't think this approach is worth it.

Reply:
True. Sorry. I badly exposed what I had in mind. What I meant is that if you opt in this attribute, either on an item or not, then you remove the possibility to have different examples and combined examples in the same docs.

Reply:
Maybe a combination of my attribute and *codeblock labels* could work. I'm just brainstorming ideas that are ideally less verbose in both scenarios. I imagine it is less common for someone to explicitly want separate scopes for every code block while **also** wanting some separation.

```rust
#[doc(test(combine_code_blocks))]
/// ```
/// let x = 0;
/// ```
/// ```
/// let y = x;
/// ```
/// ```new_scope
/// let x = 5;
/// ```
/// Below will `panic!`
/// ```
/// assert_eq!(x, 0);
/// ```
```

Maybe (and I don't know if the label syntax would support this) allowing us to set *named scopes* could work:
```rust
#[doc(test(combine_code_blocks))]
/// ```
/// let x = 0;
/// let y = x;
/// ```
/// ```set_scope(name_or_index)
/// let x = 5;
/// ```
/// Below will pass because it is part of the *default* scope
/// ```
///  assert_eq!(x, 0);
/// ```
```

This example would create two scopes for execution:
```rust
// *default* scope
{
  let x = 0;
  let y = x;
  assert_eq!(x, 0);
}

// scope(name_or_index)
{
  let x = 5;
}
```

Reply:
> Why don't we just concatenate code blocks within a single doc comment?
> It seems like most code blocks within a doc comment are in the context of the item being documented.

Yes, I had the same idea — and I think it would work very well. The stdlib examples show that it sometimes won't work, but from my experience, it's very rare to include full `main` functions in doctests, especially since the point of a doctest is to be lightweight.

My proposal would be to implement it conservatively by concatenating the code blocks per item. If this doesn't compile, then try again without concatenating.

That would allow all code to keep working.

We could perhaps remove the fallback in a new edition? I know the standard library has to be compatible across editions, but we could just rewrite the doctests to avoid the breakage.

As alternative, we could have people mark code blocks now if they want them to be concatenated and then make this a default later.